### PR TITLE
Generalize Dune Query Parameters

### PR DESCRIPTION
**Pull request by bh2smith**
Rather than having to declare dictionaries all the time with dune query parameters, I thought we could be more ergonomic and specify a type that handles string conversion of each of the 4 supported query parameters.}

While working on this, I also noticed a bug in `write_to_csv` recently introduced by removing the return statement when the data set is empty (specifically the headers are always determined by fields of the first record in the data set (an index error when the list has zero elements).

# Test Plan

Run the transfer file script and see that the CSV is as expected.

### CODE REVIEW

File: src/file_io.py [hunk 1]
```diff
@@ -20,17 +20,22 @@ def __str__(self):
...
 
+    if len(data_list) == 0:
+        # Create an empty file
+        with open(outfile.filename(), 'w', encoding='utf-8') as _:
+            pass
+        return
+
     with open(outfile.filename(), 'w', encoding='utf-8') as out_file:
+        headers = [f.name for f in fields(data_list[0])]
+        data_tuple = [astuple(x) for x in data_list]
```
Comment (MartinquaXD) [inline src/file_io.py:38]: This change should rather be a separate PR.

Also Could this be written as?:
```python
with open(outfile.filename(), 'w', encoding='utf-8') as out_file:
    if len(data_list) == 0:
        return
    headers = [f.name for f in fields(data_list[0])]
    #...
```

File: src/dune_analytics.py [hunk 2]
```diff
@@ -13,8 +17,87 @@
...
+        )
+
+    @classmethod
+    def date_type(cls, name: str, value: datetime):
+        """Constructs a Query parameter of type date"""
+        return cls(
+            name,
+            DuneParameterType.DATE,
+            raw_value=value,
+            # This is the postgres string format of timestamptz
```
Comment (MartinquaXD) [inline src/dune_analytics.py:75]: \*afraid of sounding lame 🙈 \*
`timestamptz` -> `timestamps`

File: src/dune_analytics.py [hunk 3]
```diff
@@ -13,8 +17,87 @@
...
+        )
+
+    @classmethod
+    def list_type(cls, name: str, value: list[str]):
+        """Constructs a Query parameter of type list"""
+        return cls(
+            name,
+            DuneParameterType.LIST,
+            raw_value=value,
+            value_str="\n".join(value),
```
Comment (MartinquaXD) [inline src/dune_analytics.py:86]: Using `\n` seems surprising to me. I have no idea of dune but should that maybe be a `,`?

File: src/dune_analytics.py [hunk 4]
```diff
@@ -13,8 +17,87 @@
...
+
+class QueryParameter:
+    """Class whose instances are Dune Compatible Query Parameters"""
+
+    def __init__(
+            self,
+            name: str,
+            parameter_type: DuneParameterType,
+            raw_value: Any,
+            value_str: str
```
Comment (MartinquaXD) [inline src/dune_analytics.py:41]: I think `value_str` can be dropped. It's only needed in `to_dict` and this function should have all the information it needs to correctly serialize `raw_value`.

File: src/dune_analytics.py (hunk 2 above)
Comment (bh2smith) [reply src/dune_analytics.py:75]: the primitive type in postgres is `timstamptz` timestamp with timezone

File: src/dune_analytics.py (hunk 3 above)
Comment (bh2smith) [reply src/dune_analytics.py:86]: Screenshot 2022-03-04 at 12 54 44 PM
This is the specification given by Dune.

File: src/dune_analytics.py (hunk 4 above)
Comment (bh2smith) [reply src/dune_analytics.py:41]: The problem is knowing what to serialize as (I thought it would be good to keep that in the initialization method and not have to serialize more than once. I can also move it into the `to_dict` method if you insist.

File: src/dune_analytics.py (hunk 2 above)
Comment (MartinquaXD) [reply src/dune_analytics.py:75]: Lol. Never mind then. 😅

File: src/file_io.py (hunk 1 above)
Comment (bh2smith) [reply src/file_io.py:38]: For sure - will do.

Comment (bh2smith) [reply src/file_io.py:38]: Actually, I don't like the idea of returning from within an open file. it is important the the `with open` block is completed.

File: src/dune_analytics.py (hunk 4 above)
Comment (MartinquaXD) [reply src/dune_analytics.py:41]: Seems like a code smell to have basically the same data duplicated.
One could have a switch over the `parameter_type` to know how to deserialize the value.
Alternatively one could also drop `raw_value`. So far it only gets used in `value()` and to be honest I don't expect that function to be of much use.

File: src/file_io.py (hunk 1 above)
Comment (bh2smith) [reply src/file_io.py:38]: Opened PR for this specific change #7 - we can continue the discussion there.

Comment (MartinquaXD) [reply src/file_io.py]: I really have no strong feelings about this but it seems like returning early from the `with` block would also properly close the file: https://stackoverflow.com/questions/9885217/in-python-if-i-return-inside-a-with-block-will-the-file-still-close

File: src/dune_analytics.py (hunk 4 above)
Comment (bh2smith) [reply src/dune_analytics.py:38]: Yea cool. So I will rework this right MEOW and you can let me know what you think. I totally agree that its a bit weird.

Comment (bh2smith) [reply src/dune_analytics.py]: gefixed!

File: src/file_io.py (hunk 1 above)
Comment (bh2smith) [reply src/file_io.py]: If that's the case then I will surely do this!

**Review by MartinquaXD — APPROVED** [review]
